app/lib/request.rb

```diff
@@ -202,28 +205,46 @@ def open(host, *args)
 
             begin
               sock.connect_nonblock(sockaddr)
+              # We somehow managed to connect immediately, close pending socks
+              # and return immediately
+              socks.each(&:close)
+              return sock
             rescue IO::WaitWritable
-              if IO.select(nil, [sock], nil, Request::TIMEOUT[:connect])
-                begin
-                  sock.connect_nonblock(sockaddr)
-                rescue Errno::EISCONN
-                  # Yippee!
-                rescue
-                  sock.close
-                  raise
-                end
-              else
-                sock.close
-                raise HTTP::TimeoutError, "Connect timed out after #{Request::TIMEOUT[:connect]} seconds"
-              end
+              socks << sock
+              addr_by_socket[sock] = sockaddr
```

Comment:
I am not super happy about having a socket as a hash key, but I'm not sure how else to make the sockaddr corresponding to the socket when doing the `IO.select` bit. Should be fine in practice.